WalletTool: Create design doc for WalletTool subcommands

ItoroD opened the pull request:
I have created a list of all subcommands and their options. The aim of doing this is to see what structure of the subcommands is most intuitive. Where should options be placed? what should be global and what should be local?

The structure in which we call the commands in CLI is identical to what we had before. The only change will be code changes each action/subcommand method will handle its own options as arguments passed to the method.

ItoroD commented:
I did a rough count of the usage of options in different commands. Here is what I found (see count below) So --wallet is used in all  so is --net. I think they should be global variables. Maybe password too.


>  wallettool --wallet [mywalletfile] --net [NETWORK] `dump` --dump-privkeys --dump-lookahead

I am suggesting; For every command, --wallet and --net should come before the subcommand. Might have to update my .md with this. What do you think? 

Here is an example of what dump would look like with this implemented. 

| Option                | Count |
|-----------------------|-------|
| `--wallet`            |   15  |
| `--net`               |   15  |
| `--password`          |    5  |
| `--date`              |    5  |
| `--unixtime`          |    3  |
| `--output-script-type`|    3  |
| `--force`             |    3  |
| `--pubkey`            |    2  |
| `--addr`              |    2  |
| `--dump-privkeys`     |    1  |
| `--dump-lookahead`    |    1  |
| `--output`            |    1  |
| `--payment-request`   |    1  |
| `--fee-per-vkb`       |    1  |
| `--fee-sat-per-vbyte` |    1  |
| `--locktime`          |    1  |
| `--select-addr`       |    1  |
| `--select-output`     |    1  |
| `--allow-unconfirmed` |    1  |
| `--no-pki`            |    1  |
| `--waitfor`           |    1  |

msgilligan commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,514 @@
+= Wallettool Documentation
```
I suggest "WalletTool Conversion to PicoCLI Subcommands"

The design document should describe the proposed change.

msgilligan commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,514 @@
+This document aims to highlight some issues in wallettool module and
+suggest how most of the issues can be solved and improved by migrating
+the use of `@parameter` as a way of calling various actions, to
+`Subcommands` as the first improvement step.
```
Mention PicoCLI here and provide links to the referenced `@parameter` and `Subcommands` section of the PicoCLI documentation.

msgilligan replied:
We also should use either `WalletTool` or `wallet-tool` when describing the command. Generally, I try to use `WalletTool` when referring to the code or the Java class and `wallet-tool` when referring to the binary. It might be better to have a single name that we use, but for now let's carry on with the two names and try to have consistent capitalization.

ItoroD replied:
Thanks for the feedback

msgilligan reviewed (changes requested):
This is a preliminary review. If you could make these requested changes, I'll make another pass. If you make sure I can modify the PR I will do some copyediting for you and also look in detail at the various subcommands.

I think the overall AsciiDoc format  is good and the clickable table-of-contents is nice.

msgilligan commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,579 @@
+== WalletTool: Migrate to subcommand
+
+This document aims to highlight some issues in WalletTool module and
+suggest how most of the issues can be solved and improved by migrating
```
suggest -> suggests
solved **or** improved

msgilligan commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,579 @@
+
+== WalletTool: Migrate to subcommand
+
+This document aims to highlight some issues in WalletTool module and
```
in **the** WalletTool module

msgilligan commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,579 @@
+
+This document aims to highlight some issues in WalletTool module and
+suggest how most of the issues can be solved and improved by migrating
+the use of picocli's link:https://picocli.info/#_options_and_parameters[Parameter] as a way of calling various actions, to
```
**to** the use

msgilligan commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,579 @@
+ as the first improvement step.
+
+=== Identified Problems
+The issues this migration aims to solve can be divided into 3. User Interface, Documentation, and Code Refactoring.
```
into 3 **types**:

msgilligan commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,579 @@
+The issues this migration aims to solve can be divided into 3. User Interface, Documentation, and Code Refactoring.
+
+==== User Interface
+currently the commandline print out for things like Help is difficult to read because there are lots of options mapped to wallet-tool command. This makes it difficult to tell which options go with which command/action. In addition, not all options need to be specified when there are default values for those options. If the option is not specified, we could use the defaults. This will help for a b ...
```
Currently

msgilligan commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,579 @@
+
+==== *Options*:
+
+* `--net ++<++NETWORK++>++`: Specifies the network.
```
I think this option should _only_ be used for `create` and thus only documented under the `create` subcommand.

msgilligan commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,579 @@
+
+[source,bash]
+----
+wallet-tool <WALLET_FILE> [--net <NETWORK>] <Subcommand>
```
I think the order should be:

wallet-tool  <Subcommand> [options] <WALLET_FILE>

msgilligan commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,579 @@
++
+[source,bash]
+----
+wallet-tool mywallet.wallet --net mainnet dump
```
wallet-tool dump mywallet.wallet

msgilligan commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,579 @@
++
+[source,bash]
+----
+wallet-tool newwallet.wallet create
```
wallet-tool create newwallet.wallet

msgilligan commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,579 @@
+
+[source,bash]
+----
+wallet-tool <WALLET_FILE> create [--seed <SEED>] [--watchkey <XPUB>] [--date <DATE>] [--unixtime <UNIX_TIMESTAMP>] [--output-script-type <TYPE>] [--force]
```
wallet-tool create [--net <NETWORK>] [--seed <SEED>] [--watchkey <XPUB>] [--date <DATE>] [--unixtime <UNIX_TIMESTAMP>] [--output-script-type <TYPE>] [--force]  <WALLET_FILE>

msgilligan commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,579 @@
+- `--date <DATE>`: Wallet creation date (formatted as `YYYY-MM-DD`).
+- `--unixtime <UNIX_TIMESTAMP>`: Wallet creation time in Unix timestamp format (overrides `--date` if both are specified).
+- `--output-script-type <TYPE>`: If specified, use that for deriving addresses. Provide an output script type. Valid values: P2PKH, P2WPKH. Default: P2PKH.
+- `--force`: Overwrites any existing wallet file.
```
Add --net

msgilligan replied:
I think we should default to `P2WPKH` it's the best choice we currently support.

schildbach replied:
I'd like to add to this one: when doing or describing the migration, I think we should not change any behaviour at all, also not the defaults. Refactoring is refactoring, and changes are changes – separate commits, separate PRs (in most cases)

I but strongly agree the default should be `P2WPKH`. I'm a bit embarrassed why I missed this for such a long time.

So my proposal is we change the default now, before merging the migration.

msgilligan commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,579 @@
++
+[source,bash]
+----
+wallet-tool newwallet.wallet create --seed "12 mnemonic words" --output-script-type P2WPKH
```
wallet-tool create --seed "12 mnemonic words" --output-script-type P2WPKH  newwallet.wallet

msgilligan commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,579 @@
+----
+
+==== *Options*:
+* `--output-script-type <TYPE>`: If specified, use that as the upgrade target. Script type for address derivation. Examples: `P2PKH`, `P2WPKH`.
```
Should we only allow upgrade from P2PKH to P2WPKH and not the other way?

ItoroD replied:
True, other way is not really an upgrade.

ItoroD replied:
> Yes, there is a specific upgrade path that only can be followed upwards (and without skipping).
> 
> In the moment, it's only P2PKH → P2WPKH. In future, it will likely be P2PKH → P2WPKH → P2TR.
> 
> (Upgrading from Basic wallets has been removed from bitcoinj with [4e8a199](https://github.com/bitcoinj/bitcoinj/commit/4e8a19997d4c08d3e2633cf9368d0cef9b6f7ba7))

Alright, I will take that basic out of the doc.

msgilligan commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,579 @@
+* `--select-output ++<++OUTPUT++>++`: Restricts input selection to a specific output.
+* `--waitfor ++<++CONFIRMATIONS++>++`: Waits for a specific number of confirmations.
+* `--allow-unconfirmed`: Allows you to create spends of pending non-change outputs.
+* `--no-pki`: Disables PKI verification for payment requests.
```
Payment requests have been removed (I think, if not will be removed soon) so we should remove this option.

msgilligan reviewed (changes requested):
Overall, this looks good.

I've requested some changes. Some are simply grammar/spelling.

I feel pretty strongly that the `subcommand` should precede the wallet filename as that is how most UNIX cli programs work (e.g. `git`)  I started to indicate this on every usage, but realized that would be pretty repetitive so I'll leave it to you to fix it everywhere.

Some of my suggestions for UI changes (from current behavior) might be controversial (e.g. with Andreas) but go ahead and make them and I'll try to convince him. If we have to change back to current behavior, we can change it back.

msgilligan commented:
We're getting closer. Maybe one more pass and I'll ping Andreas.

ItoroD commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,575 @@
+
+[source,bash]
+----
+wallet-tool create [--net <NETWORK>] [--seed <SEED>] [--watchkey <XPUB>] [--date <DATE>] [--unixtime <UNIX_TIMESTAMP>] [--output-script-type <TYPE>] [--force] <WALLET_FILE>
```
OR symbol should be `||` instead of `|` ? `|` means something else in .adoc

ItoroD commented:
> Looks good! Approved!
> 
> Approval means it is ready to merge to `master` as a proposal/design doc. There will still be a chance for others to provide feedback and or make change requests.

Thanks Sean!

schildbach commented on designdocs/wallettool-subcommands.adoc:
```diff
@@ -0,0 +1,575 @@
+
+=== *`upgrade`*
+
+*Description*: Upgrade basic or deterministic wallets to deterministic wallets of the given script type. If specified, uses a target script type for deriving new addresses.
+
+==== *Usage*:
+
+[source,bash]
+----
+wallet-tool upgrade [--output-script-type <TYPE>] <WALLET_FILE>
+----
+
+==== *Options*:
+* `--output-script-type <TYPE>`: If specified, use that as the upgrade target. Script type for address derivation. Note that upgrade can be made from `Basic` --> `P2PKH` --> `P2WPKH`.
```
I have to investigate, but I'm not sure if the Basic to Deterministic migration still exists. I believe not.

In this case I'll fix the *current* documentation.

schildbach replied:
Maybe related: eb15ac07d356af74216c86f969eb7f48cc032045 and 4e8a19997d4c08d3e2633cf9368d0cef9b6f7ba7

schildbach replied:
I've put up this PR: https://github.com/bitcoinj/bitcoinj/pull/3842

I lean towards merging it first (and backport to 0.17) and then adapting this PR (very simple changes).

schildbach commented:
Well written, and I agree in particular with part 1 of the document.

schildbach commented:
Found and changed two more occurences of upgrading basic wallets, which is not possible any more.

Added a LF at EOF. Generally, all text files should use LF to terminate each line, including the last one.

ItoroD commented:
> Found and changed two more occurences of upgrading basic wallets, which is not possible any more.
> 
> Added a LF at EOF. Generally, all text files should use LF to terminate each line, including the last one.



> Found and changed two more occurences of upgrading basic wallets, which is not possible any more.
> 
> Added a LF at EOF. Generally, all text files should use LF to terminate each line, including the last one.

Thanks for the catch.
